src/GafferSceneTest/TestLight.cpp

```diff
@@ -42,54 +42,86 @@
 #include "IECoreScene/Shader.h"
 
 using namespace Gaffer;
+using namespace GafferScene;
 using namespace GafferSceneTest;
 
 GAFFER_NODE_DEFINE_TYPE( TestLight )
 
+size_t TestLight::g_firstPlugIndex = 0;
+
 TestLight::TestLight( const std::string &name )
 	:	Light( name )
 {
-	addChild( new StringPlug( "lightShaderName", Plug::In, "testLight" ) );
+	storeIndexOfNextChild( g_firstPlugIndex );
+
+	addChild( new TestShader( "__shader" ) );
+	addChild( new ShaderPlug( "__shaderIn", Plug::In, Plug::Default & ~Plug::Serialisable ) );
+
+	shaderNode()->loadShader( "lightTests" );
+
+	// We need to match the plugs already in `testShader::parametersPlug()` in order to
+	// connect the `shaderNode()` parameters and our parameters. Subsequent calls to `loadShader()`
+	// will properly remove and replace our parameters because plugs are modified on the `source()`
+	// of `parametersPlug()`.
```

Comment:
I think we should be able to do this the other way around - make the connection and then call `loadShader()` at the end. Then we wouldn't need to duplicate the plugs here.

Reply:
With the current setup of having a default shader loaded in `TestShader`, I'm not sure that we can make the connection without the duplicate plugs. `addChild( new TestShader( "__shader" ) );` will cause the `simpleShader` to be loaded in the constructor. When connecting the parameters plugs, `setInput()` will check that the number of child plugs matches and fail since the `TestLight` parameters are empty.

Maybe that and the goal of removing the default after merging to `main` is enough to remove the default now?

Reply:
Ah, of course - sorry, I should have realised that. How about we do `shader()->parametersPlug()->clearChildren()` after constructing it then, and document that we can remove that hack on `main`?

Reply:
And of course on my part too, removing those plugs allows the connection. Fixed in https://github.com/GafferHQ/gaffer/pull/5846/commits/04982d06c2d59c30feef8d7e6bafcd86e15b518f